datalad/nonasyncrunner.py

```diff
@@ -0,0 +1,137 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""
+Wrapper for command and function calls, allowing for dry runs and output handling
+
+"""
+
+import logging
+import os
+import queue
+import subprocess
+import threading
+import time
+from typing import Any
+
+
+logger = logging.getLogger("datalad.runner")
+
+STDOUT_FILENO = 1
+STDERR_FILENO = 2
+
+
+class ReaderThread(threading.Thread):
+    def __init__(self, file, q):
+        super().__init__(daemon=True)
+        self.file = file
+        self.queue = q
+        self.quit = False
+
+    def __str__(self):
+        return f"ReaderThread({self.file}, {self.queue})"
+
+    def request_exit(self):
+        """
+        Request the thread to exit. This is not guaranteed to
+        have any effect, because the thread might be waiting in
+        os.read() or queue.put(). We are closing the file that read
+        is reading from here, but the queue has to be emptied in
+        another thread in order to ensure thread-exiting.
+        """
+        self.quit = True
+        self.file.close()
+
+    def run(self):
+        logger.debug("ReaderThread(%s, %s) started", self.file, self.queue)
+        while not self.quit:
+            try:
+                data = os.read(self.file.fileno(), 1024)
+            except BrokenPipeError as exc:
+                logger.debug("%s exiting (broken pipe)", self)
+                self.queue.put((self.file.fileno(), exc, time.time()))
+                return
+
+            if data == b"":
+                logger.debug("%s exiting (stream end)", self)
+                self.queue.put((self.file.fileno(), None, time.time()))
+                return
+
+            self.queue.put((self.file.fileno(), data, time.time()))
+
+
+def run_command(cmd,
+                protocol_class,
+                stdin,
+                protocol_kwargs=None,
+                **kwargs) -> Any:
+
+    catch_stdout = protocol_class.proc_out is not None
+    catch_stderr = protocol_class.proc_err is not None
+
+    kwargs = {
+        **kwargs,
+        **dict(
+            bufsize=0,
+            stdin=stdin,
+            stdout=subprocess.PIPE if catch_stdout else None,
+            stderr=subprocess.PIPE if catch_stderr else None,
+            shell=True if isinstance(cmd, str) else False
```

Comment:
`shell=True`: Not sure if we ever really want that. It's considered a security hazard, and we should always be able to properly `shlex.split()` into a list.

I think we did this somewhere before, so it's not really this PR introducing it, but I'd like to use the occasion to get rid of it.

Reply:
Please don't use `shlex.split()`, it is POSIX-only and a significant chunk of my life went into getting its use out of the command execution code. Whether or not the use of `shell=True` is a security risk should be left to the caller to decide.

Reply:
and FTR: If any splitting is to be performed, `datalad.utils.split_cmdline()` is to be used for cross-platform compatibility.

Reply:
> Cool, that's what I wanted to see.
> What I'd love to see in addition, is a `WriterThread` as well. ;-)

That was in a previous version, together with another queue (and it worked nicely to remote control an interactive python session). I removed it for now, because it seems that the existing caller of the runner only provide stdin-content when calling the runner. That seems to work for now.

If we want to change that (the queue and writer-thread can easily be re-added), we have IMHO three options:

1. [Not good] Change the runner interface to expose a pipe, queue, or similar which the caller could use to send data (That could be a major change in the interface). This would allow the main thread to send data whenever required. But the caller interface would be something like: `start process; while process is not done: write_to_process`.

2. [Ok, works with processes the generate output] The caller implements a protocol in which the receive callbacks deposit data in the writer-queue. This requires some kind of "interactive" behavior of the process, i.e. read from stdin, write something to stdout/stderr.

3. [Best, more complicated] Extend WitlessProtocol with a input callback, that is executed on the main thread and allows to provide input. That would be called whenever the input queue is not-full, therefore some flow-control primitives might also be necessary. Then the caller can decide whether to provide input to the subprocess in the input-callback or in other callbacks by putting elements into the input queue.

I think it is nice to have, but I am wondering whether there is a need for "interactive" input to the subprocess.

Reply:
> `shell=True`: Not sure if we ever really want that. It's considered a security hazard, and we should always be able to properly `shlex.split()` into a list.
> 
> I think we did this somewhere before, so it's not really this PR introducing it, but I'd like to use the occasion to get rid of it.

I kept the original semantics in this one

Reply:
> Please don't use shlex.split(), it is POSIX-only and a significant chunk of my life went into getting its use out of the command execution code.
> If any splitting is to be performed, datalad.utils.split_cmdline() is to be used for cross-platform compatibility.

True. Forgot about that.

> Whether or not the use of shell=True is a security risk should be left to the caller to decide.

Fine with me, but here we decide based on what `cmd` is an instance of. That's the "wrong" part, I think.

Reply:
> Therefore I determined, that an "interactive" stdin-content provision is not required.

The part where it is required at the moment, is the implementation of the ORA special remote. I don't go through the runner there and have an ugly construct to talk to that subprocess. Always wanted to build a proper threaded solution like this one, but never got to it.

Re 1., 2. or 3.:

The usecase with ORA is indeed an "interactive" remote shell. The special remote waits for git-annex to request something, then it goes ahead, writes a command to that remote shell's stdin, reads from stdout/stderr and proceeds according to the output. Then the shell's subprocess is "idle" until the process that runs the special remote receives the next request from annex and so on. So, an input callback sounds suboptimal at a first glance (need to think that through). The caller process needs to be able to decide, when to send the next input and must not be blocked by the subprocess waiting for an input.

Reply:
Just to emphasize: Not sure about an input callback not being a good solution. Need to have a calm moment to envision how that would look like. Not having that moment right now :)

Reply:
> Fine with me, but here we decide based on what cmd is an instance of. That's the "wrong" part, I think.

Wrong or not, this is how it is done, and that assumption is baked into the present usage. If it needs to be changed now to become "right", please book the time to assess the situation in-full, rather than in the narrow scope of what would be good for this method.

Reply:
> The usecase with ORA is indeed an "interactive" remote shell. The special remote waits for git-annex to request something, then it goes ahead, writes a command to that remote shell's stdin, reads from stdout/stderr and proceeds according to the output. Then the shell's subprocess is "idle" until the process that runs the special remote receives the next request from annex and so on. So, an input callback sounds suboptimal at a first glance (need to think that through). The caller process needs to be able to decide, when to send the next input and must not be blocked by the subprocess waiting for an input.

The current implementation passes the Popen instance of the subprocess in the connection_made-callback. That should have a pipe, if you override the stdin-keyword argument with { "stdin": subprocess.PIPE } in the kwargs to run_command.

The stdin-attribute of the Popen instance can always be used to send bytes to the stdin of the subprocess. But: only when the main thread runs in your code, which is currently (while the subprocess is running) only in the Protocol callbacks, mainly in pipe_data_received.

If that is not sufficient, we can add a provide_pipe_data-callback that is called whenever we can send data to the subprocess. This can be throttled by the size of an input queue (which holds elements that the subprocess should consume from stdin), otherwise, it would be called very often

Reply:
In the first version, no specific writer support will be implemented. It is possible to pass `subprocess.PIPE` as stdin-argument to run_command and write to this pipe, for example, in the `connection_made` or `pipe_data_received` callbacks of the protocol